joomla/joomla-cms
# [4.2] [GSoC 21] Keyboard-Shortcut Plugin

### Summary of Changes
This is a new plugin, developed in the GSoC 21 Accessibility project. 

It introduces keyboard shortcuts to the administrator site like pressing J + F to jump into the search field or J + S to save. Other extensions are able to add their shortcuts as well by hooking in on the `onLoadShortcuts` event. All(!) of them are listed in a modal window, which shows up by pressing J + X. 

The current state is just the start. Further steps would be the editing of shortcuts, the possibility to disable them, changing the overview-text-hint to a button in the corner etc. The editing was there already but it had some flaws, so it will come later (without flaws) :)

Thanks to @Krshivam25 the student behind this GSoC project, @chmst, @bembelimen, and @brianteeman for alle the work and feedback!

### Testing Instructions
Make sure the shortcut plugin is enabled and start by pressing J + X (in sequence, not at the same time!) for the overview. Have a look at the hint, which tells you the same at the bottom of the page. Then visit different pages and try out the shortcuts :)

If your motivated you can let your plugin add new shortcuts and test them! Add "onLoadShortcuts" to the getSubscribedEvents and write something like this:
```
public function onLoadShortcuts(EventInterface $event): void {
   $shortcuts = $event->getArgument('shortcuts');
   $testShortcuts = array('test' => (object)['shortcut' => 'J + T', 'title' => 'Great Test', 'selector' => '#menu-collapse']);
   $event->setArgument('shortcuts',  array_merge($shortcuts, $testShortcuts));
}
```

### Actual result BEFORE applying this Pull Request
You needed to click on buttons to do something

### Expected result AFTER applying this Pull Request
You can use your keyboard to do something (besides typing text)

### Documentation Changes Required
Yes, I will create a page to describe the build up of the shortcuts, how extensions can add them and what the default joomla shortcuts are. Part of it would be:

The 3rd party plugin needs to provide an array of shortcuts which are objects:
`{ shortcut: string, selector: string, title: string }`
The shortcut needs to be a keyboard input, separated by a plus e.g. "J + Y" or "ALT + Z + 7" (currently there is really no filtering).
Selectors are like CSS selectors to target an element. When it is an input element, the shortcut gives focus, otherwise it will be clicked.
The Title will be displayed in the overview. It could be the name of the target.
